governance/C++-Best-Practices-Guide.md

```diff
@@ -0,0 +1,462 @@
+
+#### string view printing examples
+
+```c++
+void PrintString()
+{
+    // rawString is NOT null_terminated
+    char* rawString[11] = { 'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd' };
+
+    // Constructs string_view by passing in the explicit size of the string
+    AZStd::string_view rawView(rawString, 10); // Initialized with const char* = ("Hello World" + non-initialized data after the string) and size = 10
+    printf("%s", view.data()); // Undefined Behavior: the string_view is a view into a non-null terminated string.
+    printf("%.*s", aznumeric_cast<int>(view.size()), view.data()); // Correct: only outputs characters up to the length of the string
+    printf(AZ_STRING_FORMAT, AZ_STRING_ARG(view)); // Correct: Uses MACRO: only outputs characters up to the length of the string
+
+    AZStd::string_view null_terminated_view("Hello World"); // Initialized with const char* = "Hello World" and size = 11
+    printf("%s", null_terminated_view.data()); // Technically incorrect, but Works: Requires the user to have knowledge that the string_view is viewing a   null-terminated string
+    printf("%.*s", aznumeric_cast<int>(null_terminated_view.size()), null_terminated_view.data()); // Correct: always works
+    printf(AZ_STRING_FORMAT, AZ_STRING_ARG(null_terminated_view)); // Correct: Uses MACRO always works
+
+    AZStd::string_view substr_view = null_terminated_view.substr(0, 5); // Initialized with const char* = "Hello World" and size = 5
+    printf("%s", substr_view.data()); // Incorrect: outputs the entire null_terminated string of "Hello World"
+    printf("%.*s", aznumeric_cast<int>(substr_view.size()), substr_view.data()); // Correct: outputs "Hello"
+    printf(AZ_STRING_FORMAT, AZ_STRING_ARG(substr_view)); // Correct: Uses MACRO, outputs "Hello"
+
+    AZStd::string_view empty_view; // Initialized with const char* = nullptr and size = 0
+    printf("%s", empty_view.data()); // Undefined Behavior: nullptr dereferences occurs when printf attempts to dereference the first character
+    printf("%.*s", aznumeric_cast<int>(empty_view.size()), empty_view.data()); // Correct: Will not attempt to deference the string_view as the precision specified is 0
+    printf(AZ_STRING_FORMAT, AZ_STRING_ARG(empty_view)); // Correct: Uses MACRO, outputs nothing at at all as the empty_view sets the precision to 0
+}
+```
+
+## Error Handling
+
+### Tracing
+
+**Required**: Use `AZ_*` tracing macros for all code tracing needs. For custom type tracing (tools, etc.) or something that needs to be active in release, please create a [Request For Comment (RFC)](https://github.com/o3de/sig-core/issues/new/choose) document detailing the reasons for custom tracking to facilitate feedback from other developers.
```

Comment:
In the sentence `Use 'AZ_*' tracing macros for all code tracing needs.` having just AZ_* is not useful, it'd be better to be specific and put some examples.

Reply:
Good call-out, I've updated this to be more specific, please see the latest changes. Thanks!